### PR TITLE
Adaptions to virtual builds on x86

### PR DESCRIPTION
**Pull request by noseglasses**
When build with Hardware-Virtual on x86 Kaleidoscope gcc (5.4) throws a lot of warnings due to narrowing conversions. The literal string `B00000000` is e.g. interpreted as `int` by default and then assigned to `uint8_t` in other places. Also gcc throws unused parameter warnings. It is a good idea, to generally comment any unused parameters' names to avoid such warnings that are also issued by other compilers.

### CODE REVIEW

**Comment by obra** [comment]
This PR mixes some useful stuff with stuff that I find problematic.

8c5e5ec has some straight up bug fixes. Those should ideally be submitted as their own PR. It's quick and easy to merge. But it also has an unexplained ".raw =" added in there, without documentation or explanation.

The numerous explicit casts to prevent narrowing conversion warnings that currently only trigger for a test platform build are not to my taste. They make the code less readable to me. I think I would prefer that instead the build tools not warn for Bxxxxxxx.

It looks like you've swapped IS_CONSUMER and IS_INTERNAL without comment or explanation.

Also, it looks like you've commented out the name of a variable in the definition of setup.

File: src/key_defs_sysctl.h [hunk 1]
```diff
@@ -28,7 +28,7 @@
 #define System_Break (Key) { HID_SYSTEM_BREAK, KEY_FLAGS | SYNTHETIC|IS_SYSCTL | HID_TYPE_OSC }
 #define System_DebuggerBreak (Key) { HID_SYSTEM_DEBUGGER_BREAK, KEY_FLAGS | SYNTHETIC|IS_SYSCTL | HID_TYPE_OSC }
 #define System_ApplicationBreak (Key) { HID_APPLICATION_BREAK, KEY_FLAGS | SYNTHETIC|IS_SYSCTL | HID_TYPE_OSC }
-#define System_ApplicationDebuggerBreak (Key) { HID_APPLICATION_DEBUGGER_BREAK | HID_TYPE_OSC }
+#define System_ApplicationDebuggerBreak (Key) { .raw = HID_APPLICATION_DEBUGGER_BREAK | HID_TYPE_OSC }
```
Comment (noseglasses) [inline src/key_defs_sysctl.h]: gcc (version 5.2) had problems with this line on x86. I suppose that's because there is just one value passed to the brace initializer. For the compiler it is not obvious, given the fact that we are initializing a union, to which of the union members to assign the single value (keyCode or raw). Explicitly setting `.raw` as target resolves this issue. This seems not to be a problem when two values are passed to the brace initialize list.

Arduino is using a slightly older version of gcc. That's why the error  does not show up when compiled with current Arduino's avr-gcc.

File: src/key_defs.h [hunk 2]
```diff
@@ -66,45 +66,45 @@ typedef union Key_ {
...
+#define LSHIFT(k) ((Key) { k.keyCode, uint8_t(k.flags | SHIFT_HELD) })
+#define LGUI(k)   ((Key) { k.keyCode, uint8_t(k.flags | GUI_HELD) })
 
 // we assert that synthetic keys can never have keys held, so we reuse the _HELD bits
-#define IS_SYSCTL                  B00000001
-#define IS_INTERNAL                B00000010
-#define SWITCH_TO_KEYMAP           B00000100
-#define IS_CONSUMER                B00001000
+#define IS_SYSCTL                  uint8_t(B00000001)
+#define IS_CONSUMER                uint8_t(B00000010)
```
Comment (noseglasses) [inline src/key_defs.h]: That is something, I hardly have an explanation for. Will be fixed asap.

**Comment by noseglasses** [comment]
> The numerous explicit casts to prevent narrowing conversion warnings that currently only trigger for a test platform build are not to my taste. They make the code less readable to me. I think I would prefer that instead the build tools not warn for Bxxxxxxx.

Would
```cpp
constexpr uint8_t CTRL_HELD = B00000001;
```
instead of 
```cpp
#define CTRL_HELD         uint8_t(B00000001)
```
be something you could live with? It would improve type safety and readability. This can be changed easily for all constant macros in an automatized fashion, using regex.

It would also remove the necessity to add explicit casts to some of the `Key` macro definitions.

Appart from that, I think that
```cpp
constexpr 
Key LALT(const Key &k) { return k.imposeModifier(LALT_HELD); }
```
with
```cpp
constexpr
Key Key::imposeModifier(uint8_t mod) {
   return Key(keyCode, flags | mod);
}
```
defined in the `Key` class would be a much clearer solution than collision prone macro functions like
```cpp
#define LALT(k)   ((Key) { k.keyCode, uint8_t(k.flags | LALT_HELD) })
```
You could e.g. not forget to add the appropriate amount of braces during implementation for macros that are initially not called (as happened and addressed by this PR). The `constexpr` functions ensure that the compiler sees and approves everything.

Please note that `constexpr` implies `inline`.

**Comment by noseglasses** [comment]
> Also, it looks like you've commented out the name of a variable in the definition of setup.

That's because gcc issues a warning about unused function parameters in implementations (not definitions). That's a very useful warning and it should always be turned on. On the other hand this means that deliberately unused parameters must either remain anonymous or the names must be commented, which I prefer for the sake of readability.

Arduino will evolve. Sooner or later it will be using gcc >= 5.2. So It makes sense to correct problems that manifest on other platforms right now, even if those are not targeted to end users. 

In general it is very useful, when dealing with standardized compiler languages like C++, to compile with more than one compiler (version). That hardens the implementation as it let's you find errors that one compiler detects and another doesn't. Compilers sometimes report errors that linters cannot find (unless they are compile based like [clang-tidy](http://clang.llvm.org/extra/clang-tidy/)). 

gcc and clang have evolved a lot in the recent years. Especially with respect to their verbose error and warning messages. I would clearly recommend to use different compilers to build and run a regression testbench once it is there, the more recent the compilers are, the better. This provides great benefit even for a project like Kaleidoscope that will always be used with avr-gcc.

See e.g. https://github.com/noseglasses/papageno/blob/master/.travis.yml for an example of a multi-compiler test setup.

**Comment by noseglasses** [comment]
Regarding `constexpr`, the `static` keyword has the same meaning as without `constexpr` (class functions, class data when used in class definitions or object locality for global functions (no symbol export for collision avoidance).

A class cannot define `constexpr` data members. Mind you, `constexpr` is a hint for the compile that a class or function or variable can be compile time constant. As soon as you instantiate a class that has non-static `constexpr` inventory (without declaring the instance `constexpr`), the instance will become a run time datum and actually have an inventory (event those members that were declared as `constexpr`).
```cpp
struct A {
    constexpr A() : n_(0) {}
    // constexpr int m_; // Illegal as non-static
    int n_;
};
 
int main()
{
   // This is a compile time constant
   constexpr A cA;
   //cA.n_ = 1; // Error 

   // This is a (run-time) instance
   A a;
   a.n_ = 1; // Ok
}
```

You will find a lot of `static constexpr` constants in class definitions. Since c++-11 constant definitions like
```cpp
class B {
   static constexpr uint8_t maxVal = 133;
};
```
are replacements for the less readable `enum` definitions that were found before c++-11 came along, like 
```cpp
class B {
   enum { maxVal = 133 };
};
```
Also enums did not provide a simple way to enforce the type (e.g. `uint8_t`) of a constant.

**Comment by gedankenexperimenter** [comment]
So, here's what I got out of that:
> A class cannot define `constexpr` data members
Then:
```
class B {
   static constexpr uint8_t maxVal = 133;
};
```

Lesson learned: Never ask anyone to explain the `static` keyword clusterf*ck.

**Comment by noseglasses** [comment]
> Lesson learned: Never ask anyone to explain the static keyword clusterf*ck.

My english is not good enough to detect whether that was irony. But if not, I will quit answering any general c++ related questions from now on.

**Comment by gedankenexperimenter** [comment]
It's not your fault that the horrifically overloaded `static` keyword is impossible to explain, and I do appreciate that you put some effort into answering my question, but did you not see that you made a statement and immediately contradicted that statement in the ensuing example?

**Comment by noseglasses** [comment]
> but did you not see that you made a statement and immediately contradicted that statement in the ensuing example?

No. 
```cpp
class B {
   static constexpr uint8_t maxVal = 133;
};
```
is not a data member but a compile time constant. I am not sure if the term 'data member' is precise enough. What I meant is that `maxVal` is not part of the class instance data inventory that is stored in memory.

So, let me rephrase my above statement: A class cannot define non-static constexpr non-method members.
```cpp
class B {
   constexpr uint8_t maxVal; // illegal
};
```
Does that make sense to you?

**Comment by gedankenexperimenter** [comment]
> Does that make sense to you?

Maybe… But there are two changes in the second example. The removal of `static`, and the the removal of the initialization. I'm also not so clear on the meaning of `constexpr` – I _think_ it means that it can't be modified at run-time, but I'm not sure (maybe it can't be modified after initialization, even if that's after compile time?). If it really can't be altered at run-time, then it certainly makes sense that you wouldn't _want_ to do that, because all it does is waste memory, but so much of this is non-obvious that there may be a deeper reason it's _disallowed_ that escapes me.

You've dumbed it down almost to my level – thanks!

**Comment by obra** [comment]
"Would

constexpr uint8_t CTRL_HELD = B00000001;
instead of
#define CTRL_HELD         uint8_t(B00000001)
"

Yeah, I think that's an improvement. (Though I'd want @algernon to sanity check me here.)

Ideally, that's something that should be done as a standalone pull req.

**Comment by algernon** [comment]
I agree, `constexpr uint8_t CTRL_HELD = B00000001` is an improvement over the define (and also agree about it being easier to merge as a separate PR).

**Comment by noseglasses** [comment]
There is one caveat, I feel obliged to mention. Once those macros are turned into compile time constants, they gain a scope. This would ideally be namespace `kaleidoscope` or an nested namespace that could reflect a property that the constants have in common. It is important to note, that once we introduce a namespace, it must be available everywhere the constant is used. 

When living in namespace `kaleidoscope`, those constants could be used as before from any sub-namespace of `kaleidoscope`. Code that is not living below `kaleidoscope` would either have to add `using namespace kaleidoscope;` or name the namespace explicitly `kaleidoscope::CTRL_HELD`. Just carelessly replacing pre-processor macros with compile time constants would break existing code.

Because of the pitfalls described above, I would suggest the following for a transition of any macros and function macros:
1. We turn all macros to `constexpr` compile time constants that live in an appropriate namespace. Thereby we adapt the constant names conforming to the style guide.
2. We have a second set of constants with the original names, living in global namespace but declared deprecated. Those are initialized using the constants living in the namespace.
3. After a sufficiently long time (if necessary, never) the second set of constants would be removed.

This would e.g. read as follows.
```cpp
namespace kaleidoscope {
...
constexpr uint8_t kCtrlHeld = B00000001;
...
}

// In global scope
...
__attribute__((deprecated("CTRL_HELD in global scope is deprecated. Please use kaleidoscope::kCtrlHeld instead")))
constexpr uint8_t CTRL_HELD = kaleidoscope::kCtrlHeld;
...
```

These changes do not cause any compatibility issues except for newly introduced deprecation warnings. All changes can be done easily using regex. I could prepare an individual PR if you approve.

**Comment by noseglasses** [comment]
To speed things up a little, I added another small PR https://github.com/keyboardio/Kaleidoscope/pull/271 that fixes some missing curly braces.

The issue addressed in the new PR was also addressed by this PR which probably will be closed anyway and replaced by individual PRs.

File: src/key_defs.h (hunk 2 above)
Comment (noseglasses) [reply src/key_defs.h]: I am still not sure how this happend, but it seems strangely related to this:
https://github.com/keyboardio/Kaleidoscope/commit/f57aab5f718d3b47ea2252ec6c006c96cc62eb59
